### ISSUE TITLE
`shadow_unrelated` triggers if variable is shadowed in macro

### ISSUE THREAD

**Issue by junbl** [rust-lang/rust-clippy#11423]
### Summary

When a variable is shadowed in a macro, subsequent shadowings trigger the `shadow_unrelated` lint, even if they use the value.

This is a regression from 1.70 to 1.71 (lint is not triggered in 1.70, but is in 1.71 and 1.72).

### Lint Name

shadow_unrelated

### Reproducer

I tried this code:

```rust
macro_rules! convert {
    ($v:ident) => {
        let $v = $v.to_string();
    }
}
fn main() {
    let x = 1;
    convert!(x);
    let x = x.as_str();
    println!("{x}");
}
```

I saw this happen:
Lint triggers, claiming that the initial binding is the previous one (not the binding in the macro).
```
error: `x` shadows a previous, unrelated binding
  --> src/main.rs:11:9
   |
11 |     let x = x.as_str();
   |         ^
   |
note: previous binding is here
  --> src/main.rs:9:9
   |
9  |     let x = 1;
   |         ^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#shadow_unrelated
note: the lint level is defined here
  --> src/main.rs:1:9
   |
1  | #![deny(clippy::shadow_unrelated)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^
```

I expected to see this happen:

Lint is not triggered, since the shadowed value is created using the existing value.

### Version

```text
rustc 1.71.1 (eb26296b5 2023-08-03)
binary: rustc
commit-hash: eb26296b556cef10fb713a38f3d16b9886080f26
commit-date: 2023-08-03
host: x86_64-unknown-linux-gnu
release: 1.71.1
LLVM version: 16.0.5


# cargo clippy --version                                                                             
clippy 0.1.71 (eb26296 2023-08-03)
```
